### PR TITLE
fix: forbid mint xt in certain transactions

### PR DESCRIPTION
**Pull request by mkxbl**
For these transactions: 

DepositRequest, Bonding, WithdrawPledge, WithdrawCollateral, WithdrawPledgeAndCollateral, LiquidationSignerTimeout, LiquidationUndercollateral, LiquidationFaultyWhenWarranty, LiquidationFaultyWhenRedeeming

mint xt should be forbidden

### CODE REVIEW

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 1]
```diff
@@ -132,6 +136,34 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+    match tx_type {
+        DepositRequest
+        | Bonding
+        | WithdrawPledge
+        | WithdrawCollateral
+        | WithdrawPledgeAndCollateral
+        | LiquidationSignerTimeout
+        | LiquidationUndercollateral
+        | LiquidationFaultyWhenWarranty
+        | LiquidationFaultyWhenRedeeming => forbid_mint_xt(),
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/mod.rs:150]: It is better to use `forbid_mint_xt` as default behavior, or explicitly match all types for secure.
If we add new TxType here, it is easy to forget this and leave the new type unchecked.

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 2]
```diff
@@ -132,6 +136,34 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+        _ => Ok(()),
+    }
+}
+
+fn forbid_mint_xt() -> Result<(), Error> {
+    let sudt_cell_count = QueryIter::new(load_cell_type, Source::Output)
+        .filter(|type_opt| type_opt.is_some())
+        .map(|type_opt| type_opt.unwrap())
+        .filter(|script| script.code_hash().raw_data().as_ref() == SUDT_CODE_HASH.as_ref())
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/mod.rs:159]: Should also filter args and type to match exactly the xt typescript.

File: contracts/toCKB-typescript/src/switch/mod.rs (hunk 1 above)
Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:150]: good idea

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 3]
```diff
@@ -132,6 +136,34 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+        .count();
+    if 0 != sudt_cell_count {
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/mod.rs:161]: Should we also add input check? The code now allows burn xt, which may not be vulnerable, but not expected.

Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:161]: no need, we can't prevent user from burning tokens anyhow

File: contracts/toCKB-typescript/src/switch/mod.rs (hunk 2 above)
Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:159]: It's a option, not necessary

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 4]
```diff
@@ -132,6 +136,32 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+        return Err(Error::TxInvalid);
```
Comment (cosinlink) [inline contracts/toCKB-typescript/src/switch/mod.rs:160]: It is better to add an error type to tell user 'forbid_mint_xt'

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 5]
```diff
@@ -132,6 +136,32 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+        | AuctionFaultyWhenRedeeming => Ok(()),
+        _ => forbid_mint_xt(),
+    }
+}
+
+fn forbid_mint_xt() -> Result<(), Error> {
+    let sudt_cell_count = QueryIter::new(load_cell_type, Source::Output)
+        .filter(|type_opt| type_opt.is_some())
+        .map(|type_opt| type_opt.unwrap())
+        .filter(|script| script.code_hash().raw_data().as_ref() == SUDT_CODE_HASH.as_ref())
```
Comment (cosinlink) [inline contracts/toCKB-typescript/src/switch/mod.rs:157]:
```suggestion
        .filter(|script| script.code_hash().raw_data().as_ref() == SUDT_CODE_HASH.as_ref() 
          || script.args().raw_data().as_ref() == toCKB_lock_hash )
        
```
the added code is aim to forbid user mint fake sudt token of which code_hash != SUDT_CODE_HASH and args == toCKB_lock_hash

File: contracts/toCKB-typescript/src/switch/mod.rs [hunk 6]
```diff
@@ -132,6 +136,32 @@ fn get_deletion_tx_type(data: &ToCKBCellDataView) -> Result<TxType, Error> {
...
+fn verify_xt(tx_type: &TxType) -> Result<(), Error> {
+    use TxType::*;
+    match tx_type {
+        MintXT
+        | PretermRedeem
+        | AttermRedeem
+        | AuctionSignerTimeout
+        | AuctionUnderCollateral
+        | AuctionFaultyWhenWarranty
+        | AuctionFaultyWhenRedeeming => Ok(()),
```
Comment (cosinlink) [inline contracts/toCKB-typescript/src/switch/mod.rs:148]:
```suggestion
        | PretermRedeem
        | AttermRedeem
        | AuctionSignerTimeout
        | AuctionFaultyWhenRedeeming => Ok(()),
```

for the xt token change ?

File: contracts/toCKB-typescript/src/switch/mod.rs (hunk 4 above)
Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:160]: don't want to explode error, only malicious user  would do that

File: contracts/toCKB-typescript/src/switch/mod.rs (hunk 5 above)
Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:157]: there is no need to consider fake sudt

File: contracts/toCKB-typescript/src/switch/mod.rs (hunk 6 above)
Comment (mkxbl) [reply contracts/toCKB-typescript/src/switch/mod.rs:148]: yep

**Review by huwenchao — APPROVED** [review]

**Pull request merged**